### PR TITLE
advertise specific address for a service

### PR DESCRIPTION
**Pull request by MerlinDMC**
Enable setting a specific address in a service definition for advertise.
If no specific address is given it will fallback to the node address and reassemble the old behaviour.

This does provide functionality to have consul running on a separate private interface and advertise services for a different public/private one during service discovery.

example for two services on the same node that advertise different addresses:
```json
{
  "service": {
    "name": "default-service",
    "port": 9200
  }
}

{
  "service": {
    "name": "external-service",
    "address": "10.0.1.4",
    "port": 9200
  }
}
```

consul bound to localhost via configuration file:
```json
{
  "data_dir": "./datadir-consul",
  "log_level": "INFO",
  "node_name": "0b4fe5be-1ad6-c1f2-9f9f-86caca9a67b3",
  "advertise_addr": "127.0.0.1",
  "bind_addr": "127.0.0.1",
  "server": true,
  "bootstrap": true
}
```

The result of a service discovery via DNS:

```shell
$ dig @localhost -p 8600 external-service.service.consul default-service.service.consul +answer

; <<>> DiG 9.8.3-P1 <<>> @localhost -p 8600 external-service.service.consul default-service.service.consul +answer
; (2 servers found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 27792
;; flags: qr aa rd; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 0
;; WARNING: recursion requested but not available

;; QUESTION SECTION:
;external-service.service.consul. IN	A

;; ANSWER SECTION:
external-service.service.consul. 0 IN	A	10.0.1.4

;; Query time: 0 msec
;; SERVER: 127.0.0.1#8600(127.0.0.1)
;; WHEN: Fri Jan  2 21:39:50 2015
;; MSG SIZE  rcvd: 96

;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 30860
;; flags: qr aa rd; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 0
;; WARNING: recursion requested but not available

;; QUESTION SECTION:
;default-service.service.consul.	IN	A

;; ANSWER SECTION:
default-service.service.consul.	0 IN	A	127.0.0.1

;; Query time: 0 msec
;; SERVER: 127.0.0.1#8600(127.0.0.1)
;; WHEN: Fri Jan  2 21:39:51 2015
;; MSG SIZE  rcvd: 94
```

external-service results in `10.0.1.4` because of the specific address in the service definition.
default-service results in `127.0.0.1` where the service defaults to the node address which is `127.0.0.1`.

### CODE REVIEW

**Comment by sean-** [comment]
+1

This is fantastic functionality.

File: command/agent/dns.go [hunk 1]
```diff
@@ -625,7 +629,7 @@ func (d *DNSServer) serviceSRVRecords(dc string, nodes structs.CheckServiceNodes
 		resp.Answer = append(resp.Answer, srvRec)
 
 		// Add the extra record
-		records := d.formatNodeRecord(&node.Node, srvRec.Target, dns.TypeANY, ttl)
+		records := d.formatNodeRecord(&node.Node, node.Node.Address, srvRec.Target, dns.TypeANY, ttl)
```
Comment (armon) [inline command/agent/dns.go]: I think the SRV also needs to handle the case of the override address

**Comment by armon** [comment]
Overall looks really good! I think we need to add some more tests to the DNS stuff to test the behavior works there as expected, as well as for the changes in the state_store.

**Comment by MerlinDMC** [comment]
I did add the SRV change.
If you don't add tests matching your needs I'll see to add some when I get more spare time.

**Comment by armon** [comment]
Awesome!

**Comment by armon** [comment]
Added some tests, and made a minor change to preserve API compatibility.

**Pull request merged**